staging/src/k8s.io/apiserver/plugin/pkg/audit/dynamic/dynamic.go

```diff
@@ -0,0 +1,292 @@
+		UpdateFunc: func(oldObj, newObj interface{}) {
+			manager.updateSink(oldObj.(*auditregv1alpha1.AuditSink), newObj.(*auditregv1alpha1.AuditSink))
+		},
+		DeleteFunc: func(obj interface{}) {
+			manager.deleteSink(obj.(*auditregv1alpha1.AuditSink))
+		},
+	})
+
+	return manager, nil
+}
+
+type dynamic struct {
+	// backendUpdateMutex holds an update lock on the backends
+	backendUpdateMutex   sync.Mutex
+	config               *Config
+	backends             *atomic.Value
+	webhookClientManager webhook.ClientManager
+	recorder             record.EventRecorder
+}
+
+type syncedBackends map[types.UID]*Backend
+
+// Names returns the names of the backend configurations
+func (s syncedBackends) Names() []string {
+	names := []string{}
+	for _, backend := range s {
+		names = append(names, backend.configuration.Name)
+	}
+	return names
+}
+
+// ProcessEvents proccesses the given events per current backend map
+func (d *dynamic) ProcessEvents(events ...*auditinternal.Event) {
+	for _, b := range d.GetBackends() {
+		b.delegate.ProcessEvents(events...)
+	}
+}
+
+// Run the dynamic backends
+func (d *dynamic) Run(stopCh <-chan struct{}) error {
+	// Shutdown channels are handled individually with the dynamic
+	// backend, message must be propagated out
+	go func() {
+		<-stopCh
+		d.disableUpdates()
+		d.stopAllBackends()
+	}()
+	var funcs []func() error
+	for _, b := range d.GetBackends() {
+		funcs = append(funcs, func() error {
+			return b.delegate.Run(b.stopChan)
+		})
+	}
+	return errors.AggregateGoroutines(funcs...)
+}
+
+// disableUpdates permanently disables the ability to modify backends
+// this is used to shutdown gracefully
+func (d *dynamic) disableUpdates() {
+	d.backendUpdateMutex.Lock()
+}
+
+// stopAllBackends closes the stopChan for every backend to enable
+// goroutines to terminate gracefully
+func (d *dynamic) stopAllBackends() {
+	for _, b := range d.GetBackends() {
+		close(b.stopChan)
+	}
+}
+
+// Shutdown calls the shutdown method on all backends
+func (d *dynamic) Shutdown() {
+	for _, b := range d.GetBackends() {
+		b.delegate.Shutdown()
+	}
+}
+
+// GetBackends retrieves current backends in a safe manner
+func (d *dynamic) GetBackends() syncedBackends {
+	return d.backends.Load().(syncedBackends)
+}
+
+// setBackends sets the current backends in a safe manner
+func (d *dynamic) setBackends(backends syncedBackends) {
+	d.backends.Store(backends)
+}
+
+// addSink is called by the shared informer when a sink is added
+func (d *dynamic) addSink(sink *auditregv1alpha1.AuditSink) {
+	b, err := d.createBackend(sink)
+	if err != nil {
+		msg := fmt.Sprintf("Could not add audit sink %q: %v", sink.Name, err)
+		glog.Error(msg)
+		d.recorder.Event(sink, v1.EventTypeWarning, "CreateFailed", msg)
+		return
+	}
+	backends := d.GetBackends()
+	if _, ok := backends[sink.UID]; ok {
+		glog.Errorf("Audit sink %q uid: %s already exists, could not readd", sink.Name, sink.UID)
+		return
+	}
+	backends[sink.UID] = b
+	d.backendUpdateMutex.Lock()
```

Comment:
locking after calling `d.GetBackends()` and modifying the local copy is surprising. isn't that subject to interleaving updates stomping on each other? I expected "lock, retrieve, update, store, unlock"

Reply:
yea originally had it this way, but then looked deeper into the informer. I may have groked the code wrong here, but https://github.com/pbarker/kubernetes/blob/audit-plugins/staging/src/k8s.io/client-go/tools/cache/shared_informer.go#L540 made it seem the infromer would only ever call serially. Thinking through this again though the Backend should be locked from being created or running as you mention in another comment.